quic/s2n-quic/src/provider/event/console_perf.rs

```diff
@@ -0,0 +1,449 @@
+/// #     loop {
+/// #           tokio::time::sleep(frequency).await;
+/// #           subscriber.print();
+/// #     }
+/// # });
+/// #
+/// # let server = Server::builder()
+/// #   .with_event(subscriber)?
+/// #   .start()?;
+/// #
+/// #    Ok(())
+/// # }
+///
+/// ```
+#[derive(Clone, Debug)]
+pub struct Subscriber {
+    counters: Arc<Counters>,
+    frequency: Option<Duration>,
+    format: Format,
+    spawned: bool,
+}
+
+#[non_exhaustive]
+#[allow(non_camel_case_types)]
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum Format {
+    /// Tab separated values with suffixes indicating the data units
+    TSV,
+    /// Tab separated values with no suffixes
+    ///
+    /// Rates are output as bits per second, sizes are output as number of bytes
+    /// and durations are output as microseconds
+    TSV_RAW,
+}
+
+impl Format {
+    fn print_suffix(&self) -> bool {
+        match self {
+            Self::TSV => true,
+            Self::TSV_RAW => false,
+        }
+    }
+}
+
+pub struct Builder {
+    frequency: Option<Duration>,
+    format: Format,
+}
+
+impl Default for Builder {
+    fn default() -> Self {
+        Self {
+            frequency: None,
+            format: Format::TSV,
+        }
+    }
+}
+
+impl Builder {
+    /// Sets the frequency at which performance metrics will be printed
+    ///
+    /// This uses `tokio::spawn` to spawn metric printing task,
+    /// and thus must be called within a Tokio runtime
+    ///
+    /// If frequency is not set, the caller is responsible for calling
+    /// `print_header` and `print` at the required frequency.
+    pub fn with_frequency(mut self, frequency: Duration) -> Self {
+        self.frequency = Some(frequency);
+        self
+    }
+
+    /// Sets the format that performance metrics will be printed in
+    pub fn with_format(mut self, format: Format) -> Self {
+        self.format = format;
+        self
+    }
+
+    /// Builds the [`console_perf::Subscriber`]
+    pub fn build(self) -> Subscriber {
+        Subscriber {
+            counters: Default::default(),
+            frequency: self.frequency,
+            format: self.format,
+            spawned: self.frequency.is_none(),
+        }
+    }
+}
+
+impl Subscriber {
+    /// Prints the current performance metrics to the console via STDOUT
+    pub fn print(&self) {
+        self.counters.print(self.format);
+    }
+
+    /// Prints the names of each performance metric to the console via STDOUT
+    pub fn print_header(&self) {
+        self.counters.print_header(self.format);
+    }
+
+    fn spawn(&mut self) {
+        if let (false, Some(frequency)) = (self.spawned, self.frequency) {
+            self.spawned = true;
+            self.print_header();
+            let subscriber = self.clone();
+            tokio::spawn(async move {
+                loop {
+                    tokio::time::sleep(frequency).await;
+                    subscriber.print();
+                }
+            });
```

Comment:
I'm still not entirely sold on the auto-spawning here, since it basically locks the API into using tokio permanently.

Reply:
I'll remove it, we can add it later if desired